### ISSUE TITLE
Add more custom fields or tags for additional subscriber filtering

### ISSUE THREAD

**Issue by pderksen** [davidhme/edd-drip#5]
I'm hoping for some discussion here to come up with a good solution that helps most users of this plugin.

Fields like product_name and price_name are getting recorded in Drip "Made a purchase" events, but I don't see a way to segment subscribers in Drip on these event values. Looks like they're just used for triggers.

![image](https://cloud.githubusercontent.com/assets/79667/11570755/c2eb57a8-99ad-11e5-988f-61281dbb1625.png)

Would the addition of saving these values in custom fields help, for example, to send a broadcast to all owners of a specific price_name (license) for a specific product?

We'd have to account for customers who buy multiple products and varying licenses.

Possibly a custom field `purchased_products` like WooCommerce-EDD? See https://docs.woothemes.com/document/woocommerce-drip/

Instead of an ID maybe the full product + license name, such as `Google Calendar Pro Add-on - Business License`? Not sure if this is too specific to my use case. I'm using variable prices in EDD.

Maybe tags are better here?

Would love to hear thoughts from the authors/other users.

**Comment by pderksen** [comment]
Update: Since 1.4 it looks like `price_name`is no longer getting recorded event to the purchase, yet quantity is.

![image](https://cloud.githubusercontent.com/assets/79667/11603561/757c5e6e-9a97-11e5-94c6-4d9d6d5ff95f.png)

It's possible it's due to my PR #4 though.

**Comment by davidhme** [comment]
Good point. 

I was hoping the Drip team would add the ability to segment by event properties when I first built this, but this hasn't happened yet.

I'm using an automation rule to tag customers based on which product & license level they buy.

Maybe automatically tagging people with `product_name - price_name` would be a good idea?

**Comment by davidhme** [comment]
@cssimmon any thoughts on this?

**Comment by cssimmon** [comment]
This question couldn't have come at a better time.  I also have a single product with variable pricing.  I could easily change the event to add the custom properties but I am not sure that solves the problem when you add multiple products to the mix.  Hate to complicate the issue but I ran into some similar challenges that may help further the discussion.

So over the weekend I added all the functionality needed to push the edd licensing renewal events to drip but when I finished and all was working I realized that for people with multiple products it just doesn't work.  The problem is exactly what Phil mentioned, how do you run concurrent campaigns OR the same campaign for subscribers with more than one product when all you have is a single set of properties for each subscriber.

To get more specific, when the license renewal event fires I push the subscriber product + license information to drip as custom properties (see below)

![image](https://cloud.githubusercontent.com/assets/6537251/11629193/b25a2c7e-9cc4-11e5-8d12-306043fc353e.png)

but what happens if the next day another product expires and those properties are updated.  I could add the renewal information to the event but  the event properties are not accessible except as a trigger so the campaign cannot access them unless they are saved as custom properties or tags.

I considered trying to prefix the subscriber properties with the product ID which would allow you to segment the product information but that would force you to have a campaign per product which may not be the best solution.  

Tagging is another option too but I dont think that is any better that prefixing the custom properties.

Unfortunately I dont have a solution yet but I am trying to figure one out. I will think about it more this week and will  reach out to the drip team to see if they have any suggestions.  I would like to get this resolved this week and pushed out to github so let me know if you have any ideas better than what I mentioned above.

**Comment by cssimmon** [comment]
I emailed the drip team but after thinking through the issue a bit more, I think the best option might be to prefix the custom properties with the download ID(see below). The download ID(1472) is the ID assigned in EDD. I don't want to use he product name because that is easily changed and could vary between dev and prod sites.

![image](https://cloud.githubusercontent.com/assets/6537251/11630847/54908bfc-9ccd-11e5-929e-e4a6ff38205a.png)

The downside is that this approach would require you to have a unique campaign for each product because you would need to use properties prefixed with the product ID.  The upside is that  it works for EDD users with single or multiple products AND would also handle Phil's issue because we would be able to track any number of custom properties at the product level.

I considered tagging but I feel that tags are more for segmentation, not customer level properties. Let me know if you disagree.

Hoping Rob or Derrick will respond today, I sent this link to both.

Let me know you thoughts.

**Comment by pderksen** [comment]
@cssimmon I like this approach and how it works within the confines of Drip's abilities right now. I'd definitely add a `price_name` custom field with the product ID prefix to what you have in your last screenshot of course.

Thanks for your work on this. Super helpful. I'll bug Derrick as well. :)

**Comment by derrickreimer** [comment]
Hey everyone, good discussion here. We definitely have some work to do to better support these use cases. I'll make some recommendations for workarounds and also mention features we have in the pipeline that will eventually make things even easier.

#### Use Case 1: Segmenting subscribers based on a specific product variation

For now, I would recommend using an automation rule to tag subscribers with the product/variation information like @pderksen and @davidhme mentioned above. This can be set up using liquid to pull data out of the event properties. In this case, you would set up an "Apply a tag" action with a value of `{{ event.product_name }} - {{ event.price_name }}`. If there's an ID field available in the event properties, then a more robust (but less human-readable) alternative would be to use the ID in the tag instead of a string name.

Something to look forward to: we still plan to add full segmenting based on event properties. This will be a significant engineering effort - we need to upgrade our DB version and migrate + index hundreds of millions of records :grimacing: - but I'd love to knock it out in early 2016.

#### Use Case 2: Storing/using data about multiple purchases

That seems like a reasonable workaround, @cssimmon. Since custom fields are a flat collection, they don't lend themselves well to one-to-many relationships - but it's the only tool we have right now for persisting data on a subscriber.

A possible alternative for transactional email is to use one-off emails in automation rules triggered by events that contain all the data you need in the properties. For example, if you want to send an email in direct response to a purchase, you could set up a rule triggered by "Made a purchase" and a "Send a one-off email" action that uses `{{ event.product_name }}`, etc. in the contents. But, if you need to send a data campaign with that data, this method will quickly fall down.

---

It pains me to even recommend ugly workarounds :). We are working on building out a Purchases API, which will be a native way to store purchase data for subscribers (and another natural way to segment subscribers, e.g. "has purchased X"). This could potentially be used in other creative ways, like "send subscriber campaign X in the context of purchase Y", which could solve use case 2 above. The challenge here is to keep all this from getting way too complex.

**Comment by pderksen** [comment]
Thanks @djreimer. I'll put the `{{ event.product_name }} - {{ event.price_name }}` tagging in place for now.

@cssimmon Would it be easy to add add the EDD product ID to the event properties so we can try out `{{ event.product_id }}` as well?

**Comment by cssimmon** [comment]
@pderksen Yes I will add the the prefixed product info to the subscriber properties.  I will also leave the existing properties so that I dont break anyones campaigns.

@davidhme Any issues with those changes? I can probably get all the renewal events done by the end of the week.

**Comment by davidhme** [comment]
Sounds good to me!
